src/psyclone/nemo.py

```diff
@@ -0,0 +1,954 @@
+        child_loops = walk_ast(node.content,
+                               [Block_Nonlabel_Do_Construct, Write_Stmt,
+                                Read_Stmt])
+        if child_loops:
+            # A kernel cannot contain other loops or reads or writes
+            return False
+
+        # Currently a kernel cannot contain implicit loops.
+        # TODO we may have to differentiate between implicit loops over
+        # grid points and any other implicit loop. Possibly using the
+        # scope of the array being accessed?
+        impl_loops = walk_ast(node.content, [Subscript_Triplet])
+        if impl_loops:
+            return False
+
+        return True
+
+    @property
+    def ktype(self):
+        '''
+        :returns: what type of kernel this is.
+        :rtype: str
+        '''
+        return self._kernel_type
+
+    def load(self, loop):
+        ''' Populate the state of this NemoKern object.
+
+        :param loop: node in the fparser2 AST representing a loop (explicit \
+                     or implicit).
+        :type loop: :py:class:`fparser.two.Fortran2003.Assignment_Stmt` or \
+                :py:class:`fparser.two.Fortran2003.Block_Nonlabel_Do_Construct`
+
+        :raises InternalError: if the supplied loop node is not recognised.
+        '''
+        from fparser.two.Fortran2003 import Block_Nonlabel_Do_Construct, \
+            Assignment_Stmt
+
+        if isinstance(loop, Block_Nonlabel_Do_Construct):
+            self._load_from_loop(loop)
+        elif isinstance(loop, Assignment_Stmt):
+            self._load_from_implicit_loop(loop)
+        else:
+            raise InternalError(
+                "Expecting either Block_Nonlabel_Do_Construct or "
+                "Assignment_Stmt but got {0}".format(str(type(loop))))
+
+    def _load_from_loop(self, loop):
+        '''
+        Populate the state of this NemoKern object from an fparser2
+        AST for an explicit loop.
+
+        :param loop: Node in the fparser2 AST representing an implicit loop.
+        :type loop: :py:class:`fparser.two.Fortran2003.Assignment_Stmt`
+
+        :raises InternalError: if first child of supplied loop node is not a \
+                           :py:class:`fparser.two.Fortran2003.Nonlabel_Do_Stmt`
+        '''
+        from fparser.two.Fortran2003 import Nonlabel_Do_Stmt, End_Do_Stmt
+
+        # Keep a pointer to the original loop in the AST
+        self._loop = loop
+
+        if not isinstance(loop.content[0], Nonlabel_Do_Stmt):
+            raise InternalError("Expecting Nonlabel_Do_Stmt as first child "
+                                "of Block_Nonlabel_Do_Construct but "
+                                "got {0}".format(type(loop.content[0])))
+        self._body = []
+        for content in loop.content[1:]:
+            if isinstance(content, End_Do_Stmt):
+                break
+            self._body.append(content)
+
+        #  I could get this by walking back up the tree and counting how
```

Comment:
So presumably we want to know if we have loop nesting of 2 or 3? Where would this be used and is this part of the todo list?

Reply:
I'm not sure of the relevance of this anymore - the way we analyse the code is more sophisticated than it was when I first wrote this (I think). The kernel type is in fact currently only used in the `view` method of the kernel. I've changed this to set `_kernel_type = "Explicit"` for the time being.